src/modm/math/proportional_unsigned.hpp

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2022, Thomas Sommer
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#pragma once
+
+#include <algorithm>
+#include <modm/math/utils/integer_traits.hpp>
+#include <modm/architecture/interface/assert.hpp>
+
+namespace modm {
+
+/**
+ * @brief 		Unsigned integer with arbitrary number of digits and proportional conversion
+ *
+ * @tparam D 	Number of Digits
+ *
+ * @author		Thomas Sommer
+ * @ingroup		modm_math
+ */
+template<int D>
+requires (D > 0)
+class ProportionalUnsigned {
+public:
+	static constexpr int Digits = D;
+
+	using T = uint_t<D>::least;
+	static constexpr T min = 0;
+	static constexpr T max = bitmask<D>();
+
+	constexpr ProportionalUnsigned() = default;
+
+	constexpr ProportionalUnsigned(T value) {
```

Comment:
Confirm, the implicit conversion of `ProportionalUnsigned` is way too volatile and error prone.
My last commit is an attempts to fix it:

The `ProportionalUnsigned` abstraction is needed for feature completeness of `color::HueD`: It's **hue component is a ProportionalUnsigned** because **Hue should wrap and not saturate**.

- removed the confusing implicit conversion.
- Accesing a `ProportionalUnsigned<...>' value now and only is possible with `ProportionalUnsigned<...>::value()`
- Implemented basic operators for ProportionalUnsigned
- Added a test in `color_test.cpp` to test/demonstrate hue-wrapping of `HsvD<...>`

@chris-durand  's code evaluates correct now: 
```
...
MODM_LOG_INFO << "b: " << b.value() << ", c: " << c.value() << modm::endl;

```
prints: b: 5, c: 5

Reply:
What do you think guys?

Reply:
Hi, your fix solves that case.

I very much like the idea of having a limited-precision unsigned integer type with wrapping arithmetic, but I am not convinced the proportional conversion should be implemented inside that class via non-explicit conversion constructors.

Have you thought of just having a fixed width wrapping integer class and doing the scaling with a non-member function? The first benefit is it would let the user choose the intended form of conversion (scaling, wrapping, truncating, etc.) making the type more flexible.

Secondly, doing scaling in an implicitly called constructor or assignment could make user code quite hard to read if you don't already know what is going on. If you had something like `scale_to_bits<8>(value)` in the code it would be quite apparent what happens.

Furthermore, separating the conversion gets rid of the conversion constructors where you can't return an error condition from and the irregular assignment operators returning `void`. 

Making the scaling a non-member function follows one of the STL's core principles: separating algorithms and data structures. It makes the class more generic while removing irregularities.

I have to admit I haven't thought this completely through, but it seems a very good idea to me at this point.

Reply:
Yes, its too monolithic. I'll rewrite this!

Reply:
Solid points @chris-durand.

It's too monolithic! As long as there's no further application, the scaling shouldn't even live in this limited precision type.

I'll rewrite this and put the scaling into color::Gray<..>.